In [2]:
#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
!pip install keras==2.1.6

# Import Libraries

In [3]:
import pandas as pd
import numpy as np

import warnings
warnings.simplefilter('ignore', FutureWarning)

In [4]:
import keras 

Using TensorFlow backend.


In [5]:
#import libraries for models
from keras.models import Sequential

In [6]:
# import layers 
from keras.layers import Dense

In [7]:
#train test split
from sklearn.model_selection import train_test_split

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


In [8]:
#import metrics MSE
from sklearn.metrics import mean_squared_error

# Load data in panda datafrane

In [9]:
df=pd.read_csv('concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [10]:
#column names
col_names= df.columns
col_names

Index(['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer',
       'Coarse Aggregate', 'Fine Aggregate', 'Age', 'Strength'],
      dtype='object')

# Split in predictors  and target

In [11]:
#target 
target = df['Strength']

In [12]:
#predictors
predictors =df[col_names[col_names != 'Strength']] # all columns except Strength
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

# Check if all the data is there

In [13]:
#check if all the data is there
df.isnull().count()

Cement                1030
Blast Furnace Slag    1030
Fly Ash               1030
Water                 1030
Superplasticizer      1030
Coarse Aggregate      1030
Fine Aggregate        1030
Age                   1030
Strength              1030
dtype: int64

# How many columns / features?

In [14]:
df.shape

(1030, 9)

<!-- # instantiate model -->

# compute input shape

In [15]:
#number of column / features
ncols=predictors.shape[1]
ncols

8

# Regression model function

# D: Increse hidden layers( 3 hidden layers)

In [1]:
def regression_model():

    # create model
    model = Sequential()
    # input layer with input shape of the pretoctors +  hidden layer with 10 nodes 
    model.add(Dense(10, activation='relu', input_shape=(ncols,)))
    #add two more hidden layers
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))

    model.add(Dense(1))#output
    
    # compile model usign adam optimaizer 
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# B: Normalizing data set(Predictors)



In [16]:
#normalizing predictors
predictors_normalized= ( predictors - predictors.mean() ) / predictors.std()

# function of process 1-3 combined

In [17]:
# the step 1-3
def reg_model_process_normalized():
    #Randomly split the data into a training and test sets by holding 30% of the data for testing.
    X_train_norm, X_test_norm, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.30, random_state=42)
    #build model
    model=regression_model()  
    #fit with 50 epochs
    model.fit(X_train_norm ,y_train, validation_split=0.3, epochs=50, verbose=2)
    #predict
    y_pred=model.predict(X_test_norm)
    #compute the mean squared error between the predicted concrete strength and the actual concrete strength using the mean_squared_error
    mse = mean_squared_error(y_test, y_pred)
    return mse




In [18]:
mses_normalized=[]
i=0
for i in range(50):
    mses_normalized.append(reg_model_process_normalized())







Train on 504 samples, validate on 217 samples
Epoch 1/50


2024-10-17 20:59:39.143531: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2024-10-17 20:59:39.152215: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2394300000 Hz
2024-10-17 20:59:39.152819: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55a1df77a620 executing computations on platform Host. Devices:
2024-10-17 20:59:39.152890: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2024-10-17 20:59:39.284916: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

 - 0s - loss: 1724.4137 - val_loss: 1577.1034
Epoch 2/50
 - 0s - loss: 1693.1362 - val_loss: 1551.7745
Epoch 3/50
 - 0s - loss: 1668.5333 - val_loss: 1532.3063
Epoch 4/50
 - 0s - loss: 1649.4917 - val_loss: 1516.6477
Epoch 5/50
 - 0s - loss: 1633.8003 - val_loss: 1503.8612
Epoch 6/50
 - 0s - loss: 1620.0721 - val_loss: 1492.2033
Epoch 7/50
 - 0s - loss: 1607.0961 - val_loss: 1480.5428
Epoch 8/50
 - 0s - loss: 1593.9750 - val_loss: 1468.1506
Epoch 9/50
 - 0s - loss: 1579.3163 - val_loss: 1453.4850
Epoch 10/50
 - 0s - loss: 1560.5869 - val_loss: 1433.4702
Epoch 11/50
 - 0s - loss: 1534.4414 - val_loss: 1404.1134
Epoch 12/50
 - 0s - loss: 1495.5753 - val_loss: 1361.9603
Epoch 13/50
 - 0s - loss: 1440.1693 - val_loss: 1299.7762
Epoch 14/50
 - 0s - loss: 1359.8939 - val_loss: 1214.6330
Epoch 15/50
 - 0s - loss: 1248.9166 - val_loss: 1101.0827
Epoch 16/50
 - 0s - loss: 1103.8066 - val_loss: 958.1091
Epoch 17/50
 - 0s - loss: 930.9617 - val_loss: 790.1863
Epoch 18/50
 - 0s - loss: 737.1774 - 


# How does the mean of the mean squared errors compare to that from Step B?

In [19]:
df_mses_normalized = pd.DataFrame(data=mses_normalized, columns=['MSE with three hidden layers'])


In [23]:
df_mses_normalized.head()

MSE with three hidden layers
0                    147.102817
1                    136.452116
2                    148.877474
3                    139.494542
4                    141.131901

In [21]:
# Mean of Normalized predictor MSE 
#3 hidden layers
df_mses_normalized.mean()

MSE with three hidden layers    142.646079
dtype: float64

### STEP D MSE with 3 hidden layers: 142.646079
### STEP B MSE with Normalized_Predictors:  633.157048

142 vs 633. Adding two more layers makes 4.45 times more accurate. It helps more than doubling epochs

